xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,338 @@
+            .toBuilder().set(ATTR_SERVER_ROUTING_CONFIG, config.routingConfig).build();
+    pne = InternalProtocolNegotiationEvent.withAttributes(pne, attr);
+    SslContextProviderSupplier sslContextProviderSupplier = config.sslContextProviderSupplier;
+
+    ChannelHandler handler;
+    if (sslContextProviderSupplier == null) {
+      if (fallbackProtocolNegotiator == null) {
+        ctx.fireExceptionCaught(new CertStoreException("No certificate source found!"));
+        return;
+      }
+      handler = fallbackProtocolNegotiator.newHandler(grpcHandler);
+    } else {
+      handler = new ServerSdsHandler(grpcHandler, sslContextProviderSupplier);
+    }
+    ctx.pipeline().replace(this, null, handler);
+    ctx.fireUserEventTriggered(pne);
+  }
+
+  public static final class FilterChainSelector {
+    public static final FilterChainSelector NO_FILTER_CHAIN = new FilterChainSelector(
+            Collections.<FilterChain, ServerRoutingConfig>emptyMap(), null, null);
+
+    private final Map<FilterChain, ServerRoutingConfig> routingConfigs;
+    @Nullable
+    private final SslContextProviderSupplier defaultSslContextProviderSupplier;
+    @Nullable
+    private final ServerRoutingConfig defaultRoutingConfig;
+
+    FilterChainSelector(Map<FilterChain, ServerRoutingConfig> routingConfigs,
+                        @Nullable SslContextProviderSupplier defaultSslContextProviderSupplier,
+                        @Nullable ServerRoutingConfig defaultRoutingConfig) {
+      this.routingConfigs = checkNotNull(routingConfigs, "routingConfigs");
+      this.defaultSslContextProviderSupplier = defaultSslContextProviderSupplier;
+      this.defaultRoutingConfig = defaultRoutingConfig;
+    }
+
+    @VisibleForTesting
+    Map<FilterChain, ServerRoutingConfig> getRoutingConfigs() {
+      return routingConfigs;
+    }
+
+    @VisibleForTesting
+    ServerRoutingConfig getDefaultRoutingConfig() {
+      return defaultRoutingConfig;
+    }
+
+    @VisibleForTesting
+    SslContextProviderSupplier getDefaultSslContextProviderSupplier() {
+      return defaultSslContextProviderSupplier;
+    }
+
+    /**
+     * returns null means we should close the connection.
+     */
+    @Nullable
+    public FilteredConfig select(InetSocketAddress localAddr, InetSocketAddress remoteAddr) {
+      Collection<FilterChain> filterChains = routingConfigs.keySet();
+      filterChains = filterOnDestinationPort(filterChains);
+      filterChains = filterOnIpAddress(filterChains, localAddr.getAddress(), true);
+      filterChains = filterOnServerNames(filterChains);
+      filterChains = filterOnTransportProtocol(filterChains);
+      filterChains = filterOnApplicationProtocols(filterChains);
+      filterChains =
+              filterOnSourceType(filterChains, remoteAddr.getAddress(), localAddr.getAddress());
+      filterChains = filterOnIpAddress(filterChains, remoteAddr.getAddress(), false);
+      filterChains = filterOnSourcePort(filterChains, remoteAddr.getPort());
```

Comment:
The above logic is identical to the one in `XdsClientWrapperForServerSds.getSslContextProviderSupplier(java.net.InetSocketAddress, java.net.InetSocketAddress, io.grpc.xds.EnvoyServerProtoData.Listener)` and the code is duplicated. The filter selection logic should be in one place and the SDS code should get the selected filter and the `SslContextProviderSupplier` from it. Right?

Reply:
Yes. 
Yes it is completely a copy.  We need another phase of cleanup due to refactoring.